### ISSUE TITLE
[REVIEW]: PlateReaderDataAnalyzer: a SAAS service to compute growth rates based on plate reader collected data

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#381]
**Submitting author:** @uriba (<a href="http://orcid.org/0000-0002-3097-9681">Uri Barenholz</a>)
**Repository:** <a href="https://github.com/uriba/PlateReaderDataAnalyzer" target ="_blank">https://github.com/uriba/PlateReaderDataAnalyzer</a>
**Version:** v0.0.1
**Editor:** @arfon
**Reviewer:** @pboesu
**Archive:** Pending

## Status

[![status](http://joss.theoj.org/papers/6588112e52f7bb09720ec6a3e934b4d2/status.svg)](http://joss.theoj.org/papers/6588112e52f7bb09720ec6a3e934b4d2)

Status badge code:

```
HTML: <a href="http://joss.theoj.org/papers/6588112e52f7bb09720ec6a3e934b4d2"><img src="http://joss.theoj.org/papers/6588112e52f7bb09720ec6a3e934b4d2/status.svg"></a>
Markdown: [![status](http://joss.theoj.org/papers/6588112e52f7bb09720ec6a3e934b4d2/status.svg)](http://joss.theoj.org/papers/6588112e52f7bb09720ec6a3e934b4d2)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/uriba/PlateReaderDataAnalyzer" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer questions

@pboesu, please carry out your review in this issue by updating the checklist below (please make sure you're logged in to GitHub). The reviewer guidelines are available here: http://joss.theoj.org/about#reviewer_guidelines. Any questions/concerns please let @arfon know.

### Conflict of interest

- [x] As the reviewer I confirm that there are no conflicts of interest for me to review this work (such as being a major contributor to the software).

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](http://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/uriba/PlateReaderDataAnalyzer">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [ ] **Version:** Does the release version given match the GitHub release (v0.0.1)?
- [x] **Authorship:** Has the submitting author (@uriba) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [ ] **Functionality:** Have the functional claims of the software been confirmed?
- [ ] **Performance:** Have any performance claims of the software been confirmed?

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [ ] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [ ] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [ ] **Automated tests:** Are there automated tests or manual steps described so that the function of the software can be verified?
- [ ] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Authors:** Does the `paper.md` file include a list of authors with their affiliations?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [ ] **References:** Do all archival references that should have a DOI list one (e.g., papers, datasets, software)?

**Comment by whedon** [comment]
Hello human, I'm @whedon. I'm here to help you with some common editorial tasks for JOSS. @pboesu it looks like you're currently assigned as the reviewer for this paper :tada:.

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As as reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all JOSS reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

**Comment by arfon** [comment]
@pboesu - how are you getting on here? I noticed you had updated a couple of checkboxes on the review issue but not much since.

**Comment by pboesu** [comment]
@arfon Sorry for the delay. I meant to do this last week but ended up packing my house and evacuating because of Irma instead. I just got back home today - fortunately no major damage. I'll aim to complete the review next week.

**Comment by arfon** [comment]
> @arfon Sorry for the delay. I meant to do this last week but ended up packing my house and evacuating because of Irma instead. I just got back home today - fortunately no major damage. I'll aim to complete the review next week.

Of course! Glad to hear you're safe (and the house isn't too bad). Please take your time and get to this when you can.

**Comment by pboesu** [comment]
Sorry again for the delay. Here's my first pass at this. 

## General comments
This is a mostly functional shiny app that visualizes outputs from a proprietary plate reader. I do not work with plate readers, or associated software, so I can not evaluate the claim made in the rather brief paper.md, that this application is indeed easier to use than the cited alternative tools.
##Architecturee, documentation, testing
Generally, the application lacks documentation. There is some introductory text in the README, and the About tab of the app, but since the author does not make use of typical R package infrastructure *functions are undocumented and untested*. 
I am aware that there is a distinction to be made between web applications and 'proper' R packages, but since the former can be packaged (see e.g. here https://will-landau.com/2016/11/01/appPackage/) the I would strongly prefer if @uriba would recast this as an R package, and use testing and documentation features, even if most functions are never to be invoked on the command line by the end user. This would also streamline the installation process.

Pop-up documentation is provided in the running app for some but not all drop-down lists, buttons, etc. at the very minimum these should be provided so the app user can understand the GUI.

## Functionality
The software appears incomplete, e.g. the "Plotly test" option in the "Data to display" drop down does not seem to trigger an action. Similarly, in the growth rate analysis setting the user can adjust maximum y-axis limits for some plots, but not others, and also there is no way to adjust minimum y-axis values. The growth rate analysis fails with "Error: incorrect number of dimensions" when the "Match" option is used with a single well. Such errors should be avoidable with adequate unit testing of the underlying shiny server functions.

## Software paper
I am aware that JOSS is not about verbose written descriptions. But I find two sentences rather too short. In particular the paper doesn't even mention the particular make and models of plate reader that this software is designed for.

## Other issues
mismatch in terminology: About panel talks about "plate layout file", interface has slot for "plate description file"

## Unticked list items
- Functionality: An error occurs when analysing single wells
- Performance: Unable to confirm, due to a lack of familiarity with the subject matter and alternative software solutions.
- Version: There is no github release
- Installation instructions: dependecies are stated in the readme, but the author sidesteps the R packaging process, which I would find preferable
- Functionality documentation: Is largely lacking, see general comments above.
- Automated tests: None.
- Community guidelines: None.
- References: reference `grofit` is missing a DOI

## Conclusions
Based on the above mentioned issues I recommend a *major revision*. I would really like to see this recast as an R package with appropriate documentation of both internal functions and GUI components. Ideally the author would also implement unit testing of the internal functions.

**Comment by pboesu** [comment]
@arfon Given that I feel like there's a need for a fundamental recasting of this app, is there a point in creating issues for all the points I raise?

**Comment by arfon** [comment]
> @arfon Given that I feel like there's a need for a fundamental recasting of this app, is there a point in creating issues in the target repository for all the points I raise?

Let's see what the author (@uriba) comes back with in response. I agree, it sounds like some fairly major enhancements/improvements are going to be necessary to this package before we can proceed further.

**Comment by uriba** [comment]
Dear @arfon and @pboesu,

@pboesu, many thanks for the thorough review!

I find @pboesu comments to be relevant and important to improve the quality
of the submitted software and make it more accessible for others to use
and/or contribute to.

I originally submitted this software for publication ~2.5 months ago in
order to get academic credit for work I've done during my PhD studies (but
that was not my main focus, nor of high, purely academic, importance, but
did serve, and continues to serve, fellow lab members as a useful utility),
a credit I needed for post-doc fellowship applications (which resonates
extremely well with the JOSS mission statement, as was detailed, in this blog
post <http://www.arfon.org/> by @arfon).

As the fellowships have all been submitted by now, my involvement in the
day-to-day lab work has decreased, and I currently face other obligations,
my capacity to fully address the issues raised in the review is limited.

If it is acceptable, I suggest allocating a 3 months period for me to
answer @pboesu 's comments and proceed from there. If I don't meet this
deadline, then the publication can be retracted.

I also suggest forming guidelines for the publication of SaaS services at
JOSS, as such services are different than 'classical' software packages in
terms of the technical capacity of their users (for example, not requiring
familiarity with code/installation etc.), testing methodologies,
versioning, etc. yet, given the recent advancements in cloud technology and
availability, I expect such software solutions to be increasingly adopted
as scientific open source software resources as well.

Many thanks again,

**Comment by arfon** [comment]
> If it is acceptable, I suggest allocating a 3 months period for me to answer @pboesu 's comments and proceed from there. If I don't meet this deadline, then the publication can be retracted.

@uriba - this sounds reasonable to me. Please let us know when you're ready to pick up this submission again.

**Comment by arfon** [comment]
👋@uriba - any chance you might pick this up again sometime soon?

**Comment by uriba** [comment]
Unfortunately, I don't see this happening anytime soon.
Moreover, given the advances in other web-oriented development platforms
(python+bokeh) I think better implementation options are currently
available.
This publication can therefore be dropped.
Thanks and have a nice weekend,
Uri Barenholz

**Comment by arfon** [comment]
OK thanks for letting us know @uriba.

@pboesu - thank you for all of your work on this.
